bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BVM.java

```diff
@@ -731,6 +731,9 @@ static void execute(Strand strand) {
                         } else {
                             strand.respCallback.setRefReturn(sf.refRegs[j]);
                         }
+                        if (checkIsType(sf.refRegs[j], BTypes.typeError)) {
```

Comment:
By moving into the RET we do this check for all returns. But an error return can only be possible in an RRET. This will avoid unnecessary checks for all returns.

Reply:
ahh sorry I was not clear I think, what I meant is that move the handle error logic to RET(not the typecheck). You do the type check here and keep a flag in the stack frame(ERROR return or NORMAL return), then use that flag to differentiate in RET instruction. According to the spec, error returns and normal value returns are kind of different (although we treat them same in the runtime) so I think having a flag or something to differentiate between two would be nice as well.

Reply:
This is fixed now